### ISSUE TITLE
Improper check for config directory when performing resumable upload

### ISSUE THREAD

**Issue by ddgenome** [googleapis/nodejs-storage#909]
Thank you for all your work on this package! The handling of the config directory when performing resumable uploads has a few issues.

First, [the code does not properly consider the `configPath` option when setting the `configDir` variable](https://github.com/googleapis/nodejs-storage/blob/862fb16db2990c958c0097252a44c775431a7b3f/src/file.ts#L1768).

```typescript
const configDir = xdgBasedir.config || os.tmpdir();
```

The code should probably be:

```typescript
import * as path from "path";
const configDir = (options.configPath ? path.dirname(options.configPath) : undefined) || xdgBasedir.config || os.tmpdir();
```

Second, the [check for write access to the `configDir`](https://github.com/googleapis/nodejs-storage/blob/862fb16db2990c958c0097252a44c775431a7b3f/src/file.ts#L1770) fails if the directory does not exist, leading to a ResumableUploadError.

```typescript
      fs.access(configDir, fs.constants.W_OK, err => {
        if (err) {
          if (options.resumable) {
            const error = new ResumableUploadError(
              [
                'A resumable upload could not be performed. The directory,',
                `${configDir}, is not writable. You may try another upload,`,
                'this time setting `options.resumable` to `false`.',
              ].join(' ')
            );
            stream.destroy(error);
            return;
          }
        …
        }
      }
```

It is not necessarily a problem if the config directory does not exist because the underlying implementation will create it if is does not exist, see [gcs-resumable-upload](https://github.com/googleapis/gcs-resumable-upload/blob/63199814aab127a5c3e5c1a22ac905d5b70b2209/src/index.ts#L268-L271), [configstore](https://github.com/yeoman/configstore/blob/764ac63d17de9e3d322c1c42180a8d5bec85391a/index.js#L59), and [make-dir](https://github.com/sindresorhus/make-dir#readme). It is only a problem if you do not have write access to the deepest existing parent of `configDir`.

#### Environment details

  - OS: all
  - Node.js version: all
  - npm version: all
  - `@google-cloud/storage` version: <=4.0.0

#### Steps to reproduce

Perform an upload with resumable set to true after removing/hiding ~/.config and you get a ResumableUploadError. Here is some example code you can use to demonstrate the issue after replacing `YOUR_BUCKET` as appropriate.

```typescript
import { Storage } from "@google-cloud/storage";
async function main(): Promise<void> {
    const storage = new Storage();
    await storage.bucket("YOUR_BUCKET")
        .upload(__filename, { resumable: true });
}
main().catch(e => {
    process.stderr.write(`Error: ${e.message}\n`);
    process.exit(1);
});
```

On a POSIX-like OS, e.g., GNU/Linux or macOS, you save the above script as "test.ts", compile it to JavaScript, and run it while hiding your `$HOME/.config` with the following command:

```console
$ HOME=/tmp GOOGLE_APPLICATION_CREDENTIALS=path/to/creds node test.js
Error: A resumable upload could not be performed. The directory, /tmp/.config, is not writable. You may try another upload, this time setting `options.resumable` to `false`.
```

This fails even though you would have permissions to create `/tmp/.config`. In other words, this succeeds:

```console
$ mkdir /tmp/.config
$ HOME=/tmp GOOGLE_APPLICATION_CREDENTIALS=path/to/creds node test.js
```

You can verify that `configPath` is _not_ checked and is created if it does not exist by changing the above example to explicitly set the `configPath` option to a non-existing directory and verifying the directory exists after running the script.

```typescript
import { Storage } from "@google-cloud/storage";
async function main(): Promise<void> {
    const storage = new Storage();
    await storage.bucket("YOUR_BUCKET")
        .upload(__filename, { configPath: "/tmp/does/not/exist", resumable: true });
}
main().catch(e => {
    process.stderr.write(`Error: ${e.message}\n`);
    process.exit(1);
});
```

```console
$ GOOGLE_APPLICATION_CREDENTIALS=path/to/creds node test.js
$ ls -al /tmp/does/not/exist
-rw------- 1 XX YY 67 Oct 30 08:17 /tmp/does/not/exist
```

Fixing the writable check will be challenging since you would have to parse the path and check `fs.constants.W_OK` access on the `configDir` and each of its parents in turn, discarding results for paths that do not exist, and stopping once you found one that does exist and only consider that result in the test. It may just be better to let the downstream packages handle the work and properly respond to errors from them.

**Comment by stephenplusplus** [comment]
Hey @ddgenome! Just a heads up I put a PR to address this issue. Please take a look when you get a chance! #915

**Comment by jkwlui** [comment]
Fixed in @google-cloud/storage@4.1.2 release

**Comment by mgabeler-lee-6rs** [comment]
> Second, the check for write access to the configDir fails if the directory does not exist, leading to a ResumableUploadError

This portion of this issue was never fixed. Opened #1420 for that
